downstairs/src/region.rs

```diff
@@ -1093,23 +1121,41 @@ impl Extent {
         cdt::extent__write__sqlite__insert__start!(|| {
             (job_id, self.number, writes.len() as u64)
         });
-        let tx = inner.metadb_transaction()?;
+        let tx = inner.metadb.transaction()?;
         for write in writes {
             if writes_to_skip.contains(&write.offset.value) {
                 debug_assert!(only_write_unwritten);
                 continue;
             }
 
+            // TODO it would be nice if we could profile what % of time we're
+            // spending on hashes locally vs sqlite
+            let on_disk_hash = integrity_hash(&[&write.data[..]]);
+
             Inner::tx_set_block_context(
                 &tx,
                 &DownstairsBlockContext {
                     block_context: write.block_context.clone(),
                     block: write.offset.value,
-                    on_disk_hash: integrity_hash(&[&write.data[..]]),
+                    on_disk_hash,
                 },
             )?;
+
+            // Worth some thought: this could happen inside
+            // tx_set_block_context, if we passed a reference to dirty_blocks
+            // into that function too. This might be nice, since then a block
+            // context could never be set without marking the block as dirty.
+            // On the other paw, our test suite very much likes the fact that
+            // tx_set_block_context would mark blocks as dirty, since it lets
+            // us easily test specific edge-cases of the database state.
+            // Could make another function that wraps tx_set_block_context
+            // and handles this as well.
+            let _ = inner
+                .dirty_blocks
+                .insert(write.offset.value as usize, on_disk_hash);
```

Comment:
I think there's a consistency problem here: if any of the `write_all` calls below return Err, the value in `dirty_blocks` at a particular offset will be wrong. If the Upstairs then sends a Flush, we'll be removing the wrong context rows.

Reply:
I think you're right about that, let me adjust. I wonder if there's some way I can force an error to test for this situation.